rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+            // let tz = Some(Arc::new(arr.timezone().unwrap()));
+            let tz = Some("UTC".into());
+
+            let min_value_timestamp = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            let max_value_timestamp = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value =
+                ScalarValue::TimestampMicrosecond(Some(min_value_timestamp), tz.clone());
+            scalar_max_value =
+                ScalarValue::TimestampMicrosecond(Some(max_value_timestamp), tz.clone());
+        }
+        DataType::Duration(_) => {
+            let min_value_duration = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            let max_value_duration = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value = ScalarValue::DurationMillisecond(Some(min_value_duration));
+            scalar_max_value = ScalarValue::DurationMillisecond(Some(max_value_duration));
+        }
+        _ => {}
+    }
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: scalar_min_value,
+        max_value: scalar_max_value,
+    }
+}
+
+fn get_decimal_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    // TODO: should this be done with Decimal128 and not i128?
+    let mut min_value = i128::MAX;
+    let mut max_value = i128::MIN;
+    let mut null_count: i64 = 0;
+
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<Decimal128Array>().unwrap())
+        .collect::<Vec<_>>();
+
+    for array in arr.iter() {
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                if let Some(Ordering::Greater) = value.partial_cmp(&max_value) {
+                    max_value = value;
+                } else if let Some(Ordering::Less) = value.partial_cmp(&min_value) {
+                    min_value = value;
+                }
+            };
+        });
+        null_count += array.null_count() as i64;
+    }
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::Decimal128(Some(min_value), 8, 8),
+        max_value: ScalarValue::Decimal128(Some(max_value), 8, 8),
+    }
+}
+
+/// Truncate a UTF8 slice to the longest prefix that is still a valid UTF8 string, while being less than `length` bytes.
+fn truncate_utf8(data: &str, length: usize) -> Option<Vec<u8>> {
```

Comment:
We shouldn't need to allocate and, since we promise the result is valid UTF-8, we can return `&str`.
```suggestion
fn truncate_utf8(data: &str, length: usize) -> Option<&str> {
```